### DISCUSSION TITLE
textTheme and primaryTextTheme

### DISCUSSION THREAD

**Discussion by rddewan** [rydmike/flex_color_scheme#142] [Q&A] [answered]
what does it actually mean by below code? can any one give a simple and clear answer 
```
    /// Text with a color that contrasts with the card and canvas colors.
    final TextTheme? textTheme,
```

and 
```
    /// A text theme that contrasts with the primary color.
    final TextTheme? primaryTextTheme,
```

**Comment by rydmike** [comment, marked as answer]
Not sure I'm the best at simple answers, well at least not short ones 😄 

Here is both a TLDR (trying a new thing for me here, lol) and comprehensive answer.

## Case `textTheme`

For `textTheme` it means that the text has a color that is clearly readable on the background colors `ThemeData.cardColor` and `ThemeData.canvasColor`. This is done using a contrasting color that is clearly visible on those background colors.

**TLDR**
In simple terms it just means:

* white/light background -> black/dark text 
* black/dark background -> white/light text 


It is not entirely that simple though.

In current theming, in most `ThemeData` factory constructors the `ThemeData.cardColor` is assigned the value of `ThemeData.colorScheme.surface` and `ThemeData.canvasColor` the value of `ThemeData.colorScheme.surface`.

> **NOTE:** The colors `ThemeData.cardColor` and `ThemeData.canvasColor` are legacy colors and are on the way to be deprecated, so then the `ThemeData.colorScheme` will be used directly. See https://github.com/flutter/flutter/issues/91772

As for what colors those actually are, well that obviously depends on your `ColorScheme` for the theme.

Typically in a light **M2** theme, it is an almost black grey, and in dark M2 theme it is actually plain white. 

In an M3 theme, the `colorScheme.surface` is used as background and `colorScheme.onSurface` as contrast color. In a M3 color system created theme, the `onSurface` uses surface tonal palette tone 10 in light mode and tone 90 in dark mode, so the contrast colors are not plain black/white or even dark-grey/light-grey. They are white/black with hint of the primary color in them with very low chroma of primary color used, so still almost more like light and dark grey, but with a hint of primary color in them. It is subtle but the primary color influences the onSurface colors.

The `textTheme` (and `primaryTextTheme`) may also uses opacities on some of its styles using the contrast color as base, making some of its styles appear grey or tinted by the background color they are on. The opacities vary by `TextStyle` in the `TextTheme`, not all have it, and it is different in light and dark mode.

Additionally, this opacity is only used by older Typography 2014 (Flutter none M3 mode default ThemeData) and 2018 (M2 the Typography an M2 app should use, but does not by default). The new 2021 (M3 Typography, used by default when opting on on M3) does not use any opacity on `TextStyle` colors in its `TextTheme`, instead the hint of color comes from the mentioned `onSurface` color. Since there is no opacity and once the color is computed in the theme, using this might be a bit less costly in layouts. The algorithm used to compute all the colors is expensive, but is only used when you define/change the theme.

## Case `primaryTextTheme`

The idea here is the same. The `primaryTextTheme` will get a color that is readable on used `primaryColor`. 

However, since primary Color is typically not a white (or close to it) or black (or close to it) background, it can be any color, an algorithm is used to calculate if black or white color produces more contrast when it is drawn on the `primaryColor`, and it uses the color that has better contrast. (The used function is `ThemeData.estimateBrightnessForColor`).

It is important to recognize that the `primaryColor` referred to, is `ThemeData.primaryColor` not `ThemeData.colorScheme.primary`. These two colors are by default the same only in light theme mode, in dark theme mode they are not. Which is very confusing. 

In dark theme mode, the `ThemeData.primaryColor` is actually reassigned to be equal to `ThemeData.colorScheme.surface` thus being equal to `ThemeData.cardColor`, so in dark theme mode, the color of `primaryTextTheme` actually is the same color as `textTheme`. 

> **LEGACY:** The reason why this strange reassignment happens is old legacy having to do with the color of the `AppBar` needing to be dark in background color in dark mode and not primary color, and this silly reassignment was a fix for dark mode, when there was no other way to get this result. This old "hack" is no longer used by Flutter and its `AppBar`, yet this confusing color reassignment of `primaryColor` in dark mode still exists.

What this leads to is of course that the `primaryTextTheme` works great in light theme mode to put text on any container/background using color `ThemeData.primaryColor` or `ThemeData.colorScheme.primary`, but in dark theme it does not necessarily work on `ThemeData.colorScheme.primary`, unless `colorScheme.primary` happens to be a color that is `Brightness.dark` and requires contrast `Brightness.light`, which it never is an M3 based theme using the M3 Color System. In hand made M2 (M3 too) themes, it can of course sometimes be so.

This situation is imo very messed up. So FCS does not do this `ThemeData.primaryColor` reassignment, there is no need for it in Flutter anymore. As long as `ThemeData.primaryColor` exists (it will also be deprecated) FCS keeps `Themedata.primaryColor` equal to `ThemeData.colorScheme.primary` and thus the text theme named `primaryTextTheme` also always works in dark mode when you put it on "primary" color, be it the one in `ThemeData` directly or in `ThemeData.colorScheme`.

I opened an issue about this because it is very confusing for most devs. If it is of interest, you can find it here: https://github.com/flutter/flutter/issues/118146
